### ISSUE TITLE
Advance HZ_ASSERT and HZ_CORE_ASSERT macros to take in no arguments except condition

### ISSUE THREAD

**Issue by alohaeee** [TheCherno/Hazel#289]
Current asserts macros could't provide a call without user message. 
![image](https://user-images.githubusercontent.com/45694435/88662081-20d15f00-d0e2-11ea-93ea-0476b98a722b.png)
It's noticed in TODO before macros definition: "// TODO: Make this macro able to take in no arguments except condition".

Current implementation also can't help with finding a assert call with no-printable argument. 
![image](https://user-images.githubusercontent.com/45694435/88663034-90941980-d0e3-11ea-8ecc-222b1bf197b9.png)
As in the example above the compiler will note only internal format library error but not a problem call.

### Suggestion
Replace the current assert definition
```cpp
#define HZ_ASSERT(x, ...) { if(!(x)) { HZ_ERROR("Assertion Failed: {0}", __VA_ARGS__); HZ_DEBUGBREAK(); } }
#define HZ_CORE_ASSERT(x, ...) { if(!(x)) { HZ_CORE_ERROR("Assertion Failed: {0}", __VA_ARGS__); HZ_DEBUGBREAK(); } }
```
to new one:
```cpp
#define HZ_ASSERT(x, ...) { if(!(x)) { Hazel::_AssertMessage(__FILE__, __LINE__, __VA_ARGS__); HZ_DEBUGBREAK(); } }
#define HZ_CORE_ASSERT(x, ...) { if(!(x)) { Hazel::_CoreAssertMessage(__FILE__, __LINE__, __VA_ARGS__); HZ_DEBUGBREAK(); } } 
```
Provide a separate file for assert log functions:
_Assert.h_
```cpp
#pragma once

namespace Hazel
{
	void _AssertMessage(const char* file, int line);
					   
	void _AssertMessage(const char* file, int line, const char* msg);

	void _CoreAssertMessage(const char *file, int line);

	void _CoreAssertMessage(const char* file, int line, const char* msg);

}
```
_Assert.cpp_
```cpp
#include "hzpch.h"
#include "Hazel/Core/Assert.h"

#include "Hazel/Core/Log.h"

namespace Hazel
{
	void AssertMessage(const char* file, int line)
	{
		HZ_ERROR("Assertion Failed at file: {0} Line: {1}", file, line);
	}

	void _AssertMessage([[maybe_unused]] const char* file,
		[[maybe_unused]] int line, const char* msg)
	{
		HZ_ERROR("Assertion Failed: {0}", msg);
	}

	void _CoreAssertMessage(const char* file, int line)
	{
		HZ_CORE_ERROR("Assertion Failed at file: {0} Line: {1}", file, line);
	}

	void _CoreAssertMessage([[maybe_unused]] const char* file, 
		[[maybe_unused]] int line, const char* msg)
	{
		HZ_CORE_ERROR("Assertion Failed: {0}", msg);
	}
}

```
#### Pros
As result:
 - we can call assert with no arguments except expression-condition
![image](https://user-images.githubusercontent.com/45694435/88664998-9a6b4c00-d0e6-11ea-8b0d-d691dafd6b92.png)

- doesn't break the old assert code:
![image](https://user-images.githubusercontent.com/45694435/88665190-eae2a980-d0e6-11ea-9805-147334ca99ef.png)

- checking for non-string objects
![image](https://user-images.githubusercontent.com/45694435/88666109-495c5780-d0e8-11ea-8480-8bfbcff47100.png)
- checking arguments count. Passing more than 2 arguments is not more valid.

#### Cons
 - Obejects with overrided operator '<<' can't be passed as argument anymore. But we can use fmt::format for that purposes.
 - Change const char* to std::string_view in log functions later will allow to pass std::string without c_str(). But for now std::string_view is not using in Hazel.

**Comment by LovelySanta** [comment]
I do remember cherno mentioning in that video that Hazel-dev is using a macro-only solution for the problem of not providing a message. I'm not sure how much thought we should put into allowing objects to be passed in. It seems like the build in error should be sufficient in that case?

**Comment by alohaeee** [comment]
@LovelySanta , I missed that Hazel-dev already have solution for issued problem. And I can't understand what do you exactly mean about allowing object to be passed in?

**Comment by LovelySanta** [comment]
> Hazel-dev already have solution for issued problem

I mean with Hazel-dev having a macro-only solution meaning that solution might be better than another function call to `AssertMessage(...)` as you proposed. It's logging after all, so we want to keep the actual code to a minimum such that it will not fail to log messages.

> I can't understand what do you exactly mean about allowing object to be passed in?

I was referring to your comment *"Current implementation also can't help with finding a assert call with no-printable argument."* where you tested it with Test{ 1, 2 }. The message the compiler provides seems good enough in that aspect, rather than your suggestion where it errors about not having an overload function.

**Comment by alohaeee** [comment]
@LovelySanta  ty for reply, now got it. 

>  I mean with Hazel-dev having a macro-only solution meaning macro that solution might be better than another function call to AssertMessage(...) as you proposed

1) that function will be called maximum once at debug mode. This is assertion logging but not per-frame called function, where one more call matter.
2) macro-only solution is not really fair argument because loging macro will be replaced with loging function anyway. So small new wrapper functions will not greatly affect performance or whatever else.

> So we want to keep the actual code to a minimum

To keep actual code to minimum can help forcing the compiler to inline these new functions. 

>  it will not fail to log messages

I can't mention how function wrapper can be cause of fails to log messages.

> The message the compiler provides seems good enough in that aspect, rather than your suggestion where it errors about not having an overload function.

The main problem of current compiler's message is that it's doesn't points to a problem call at all. You  will only know about internal format library message as i said and show at screens before. It points to fmt code, not to code where you call assertion with wrong argument. As usual, it will be difficult to find a wrong call later after several hundred lines of code. In my implementation compiler points to wrong call location that helps a lot with actually problem finding. 

And again, current implementation allows that kind of calling:
```cpp
HZ_ASSERT(1+1 == 1, "No" , "1, "+1" , "=2");
```

Anyway, for now we can't talk about Hazel-dev implementation of that macros. Maybe it solves that problems, maybe not. So for now we can leave this issues until Hazel-dev assert macro will appear.

**Comment by LovelySanta** [comment]
After some research I came across [https://stackoverflow.com/questions/5355241/generating-function-declaration-using-a-macro-iteration/5355946#5355946](this stack overflow post). The gists about it is that it selects a correct macro depending on the amount of arguments that have been provided:
```cpp
#define PP_HAS_ARGS_SOURCE() MULTI, MULTI, ONE, ERROR
```
So we could do something similar? Like make a macro that expands the `__VA_ARGS__` to the point where it will always have a first argument, and sometimes it will have a second (and potentially more) argument (second argument is the message). By appending 2 macro names like `HZ_ASSERT_INTERNAL_WITH_MSG` and `HZ_ASSERT_INTERNAL_NO_MSG`, we can extract the correct macro name:
- In case there is only 1 argument passed (so no message), the 2nd argument passed would be empty. This means the 2nd argument will evaluate `HZ_ASSERT_INTERNAL_WITH_MSG` and the 3rd argument will be `HZ_ASSERT_INTERNAL_NO_MSG` and such we return the 3rd argument as macro name.
- In case there is more than 1 argument passed (so a message, with potential extra arguments), the 2nd argument will be filled. Assume we filter only the 1st and 2nd argument, leaving out the excess ones, the 3rd argument will be `HZ_ASSERT_INTERNAL_WITH_MSG` and the 4th argument will be `HZ_ASSERT_INTERNAL_NO_MSG` and such we return the 3rd argument as macro name.

If I recall cherno did something similar in Hazel-dev, but I recall having some issues with the `__VA_ARGS__` which seems something to be only happening in MSVC and with no other compilers...

**Comment by LovelySanta** [comment]
This is resolved in #306, I added it to the initial comment there as well.
